### PR TITLE
[5.2] Category Item count takes publish dates into account

### PR DESCRIPTION
**Pull request by hendrikbehncke**
### Summary of Changes
Item count on a category takes the publish_up and publish_down dates into account.

### Actual result BEFORE applying this Pull Request
When setting a publish_down date in the past to unpublish an article the item count on a category will still count this article as published. And vice versa with  a publish_up date in the future. The Categories view shows a different amount of articles than actually visible on the Category Blog view

### Expected result AFTER applying this Pull Request
Categories view will show the exact number of articles published in a category.

### CODE REVIEW

**Comment by richard67** [comment]
Can we be really sure that all tables `$this->_table` used here https://github.com/joomla/joomla-cms/blob/7c3d33293f4ea611955da872282c091cfa3f729c/libraries/src/Categories/Categories.php#L341 have `publish_up` and `publish_down` columns? If that table doesn't have these columns, we will get an SQL error at that place.

**Comment by Denitz** [comment]
It will frustrate a user: you will see X published articles but if you click the red counter and navigate to article list with `&filter[published]=0` filter you will see no articles because they are actually published (`#__content.state = 1`).

Not a good idea. Either skip it or implement a separate counter of smth like "currently unpublished" items.

**Comment by hendrikbehncke** [comment]
Hi @richard67, that could indeed be an issue. 
Maybe passing the fields in the options in `components/com_content/src/Service/Category.php` could solve this?

Or alternatively people use 3rd party plugins to automatically update the `published` flag depending on the `publish_up` and `publish_down` dates.

**Comment by HLeithner** [comment]
This pull request has been automatically rebased to 4.4-dev.

**Comment by HLeithner** [comment]
This pull request has been automatically rebased to 5.2-dev.

**Comment by Hackwar** [comment]
As @richard67 wrote, it isn't guaranteed that each table has these publish_up and publish_down columns, so you would have to definitely check for their existence. Since this is not exactly possible the way this is currently implemented, maybe change the parameter for countitems to allow this? For example if it is set to 2, it is checking for those fields and otherwise it will have the old behavior? You would then have to change this in the core where applicable.

**Comment by Hackwar** [comment]
@hendrikbehncke Can you please work on the changes I asked for? Thank you!

**Comment by rdeutz** [comment]
Here we have more a wording than an implementation problem. We have a state, can be published, unplublished, archived and trashed. And we have a visibility based on a timestamp, that we call it "Start Published" and "Finish Publshed" is misleading. So the category view shows the published items and includes the non visible items. The implementation here has also the problem that the not visible items are not counted in under the unpublished.

Long story short, we agreed to close this PR.

**Comment by brianteeman** [comment]
It was incorrect to close this PR as it is confusing to SITE visitors to see an article count of 5 but only see 4 articles due to the item being expired or not published yet - it is not about the count of published items in the admin UI. 

The changes to check that publish_up exists should have been implemented as suggested by various comments.
